File: data/Dtos/State.cs
using data.Entities;
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations.Schema;
using System.ComponentModel.DataAnnotations;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using data.Utility.Paging;

namespace data.Dtos
{
    public class StateRes
    {
        public int StateId { get; set; }
        public string? Name { get; set; }
        public string? Code { get; set; }

        public int CountryId { get; set; }
        public CountryRes? Country { get; set; }
    }

    public class StateReqEdit
    {
        [Required, MaxLength(255)]
        public string? Name { get; set; }
        [Required, MaxLength(10)]
        public string? Code { get; set; }

        // Foreign keys
        [Required]
        public int CountryId { get; set; }
    }

    public class StateReqSearch : PagedRequestDto
    {
        public int CountryId { get; set; }
    }

    public class StateResWithCitiesCount : StateRes
    {
        public int CitiesCount { get; set; }
    }

    public class StateResWithCountryAndCitiesCount : StateRes
    {
        public string? CountryName { get; set; }
        public int CitiesCount { get; set; }
    }
}

File: web/src/dtos/State.ts
import Common from "../utility/Common";
import { CountryRes } from "./Country";
import { PagedReq } from "./PagedReq";

export interface StateRes {
  stateId?: number;
  name?: string;
  code?: string;
  countryId?: number;
  country?: CountryRes;
}

export interface StateResWithCitiesCount extends StateRes {
  citiesCount?: number;
}

export interface StateResWithCountryAndCitiesCount extends StateRes {
  citiesCount?: number;
  countryName?: string;
}

export class StateReqEdit {
  name?: string = "";
  code?: string = "";
  countryId?: number = 0;
  constructor(countryId?: number) {
    this.countryId = countryId;
  }
}

export class StateReqSearch extends PagedReq {
  countryId?: number;

  constructor(
    {
      pageNumber = 1,
      pageSize = Common.DEFAULT_PAGE_SIZE,
      orderBy,
      searchText = "",
    }: PagedReq,
    countryId = 0
  ) {
    super({
      pageNumber: pageNumber,
      pageSize: pageSize,
      orderBy: orderBy,
      searchText: searchText,
    });
    this.countryId = countryId;
  }
}
